src/locktower.rs

```diff
@@ -0,0 +1,172 @@
+    /// without 66.7% total network commitment to that fork.
+    ///
+    /// It is not important for the rest of the network to also have the same lockout on the fork.
+    /// The worst case is that the lockout for this node will double to 512 due to
+    /// rollback for the rest of the network.  So any lockout of network commitment is acceptable,
+    /// but it must cover the required threshold percentage of network stake.
+    converge_depth: usize,
+    converge_amount: f64,
+    /// The last finalized fork.  This value is initialized at the start with some pre-defined
+    /// network value, such as fork 0.
+    finalized_fork: u64,
+}
+
+/// This should be implemented by the Bank
+pub trait ForkTree {
+    /// Return true if the leaf is a descendant of the trunk.
+    fn is_trunk_of(&self, trunk: u64, leaf: u64) -> bool;
+    /// Return what percentage of the network has agreed to the fork or a descendant fork.
+    fn convergance(&self, fork: u64) -> f64;
+}
+
+impl LockTower {
+    pub fn new(
+        max_size: usize,
+        converge_depth: usize,
+        converge_amount: f64,
+        finalized_fork: u64,
+    ) -> Self {
+        LockTower {
+            vote_locks: VecDeque::new(),
+            max_size,
+            converge_depth,
+            converge_amount,
+            finalized_fork,
+        }
+    }
+
+    /// * vote - fork id to vote on
+    /// Returns true if this vote is unlocked.
+    pub fn is_unlocked<T: ForkTree>(&self, vote: u64, tree: &T) -> bool {
+        let ix = self.num_cleared_votes(vote);
+        self.vote_locks
+            .get(ix)
+            .map(|lockout| tree.is_trunk_of(lockout.fork, vote))
+            .unwrap_or(tree.is_trunk_of(self.finalized_fork, vote))
+    }
+
+    /// * vote - fork id to vote on
+    /// Returns true if this vote is converging at the configured converge_depth
+    /// and converge_amount.
+    pub fn is_converging<T: ForkTree>(&self, vote: u64, tree: &T) -> bool {
+        let ix = self.num_cleared_votes(vote);
+        let depth = ix + self.converge_depth;
+        self.vote_locks
+            .get(depth)
+            .map(|lockout| self.converge_amount < tree.convergance(lockout.fork))
+            .unwrap_or(true)
+    }
+
+    /// * vote - fork id to vote on
+    /// Total lockout of this vote.  This function simulates the generated lockout if this vote was
+    /// applied.
+    /// The fork selection algorithm should compute the vote_weight for everyone in the network for
+    /// this vote and pick the highest stake weighted option.
+    pub fn vote_weight<T: ForkTree>(&mut self, vote: u64, tree: &T) -> u64 {
+        if !self.is_converging(vote, tree) {
+            return 0;
+        }
+        if !self.is_unlocked(vote, tree) {
+            return 0;
+        }
+        // skip any expired votes
+        let ix = self.num_cleared_votes(vote);
+        let mut prev = 2;
+        let mut total = 2;
+        for ix in ix..self.vote_locks.len() {
```

Comment:
terrible, double use of `ix` with two different meanings, neither of which is `instruction`